# Chapter 9: 迴歸分析1-類別型變數
由於列聯表分析只能探討中風與一種解釋變數之間的關係，迴歸分析則可以提供我們一次分析多種解釋變數對於中風的影響。
在這個部份，我們首先做了只有類別型變數的迴歸模型，下一個部份則做了只有數值型變數的迴歸模型，第三部份結合了所有的變數並做出一個最終模型。  

<span style="color:blue"> *從前面列聯表分析中: 中風與"性別"及"住處型態"無關，因此在迴歸分析是直接被拿掉的*</span>

## 讀取資料

In [1]:
strokedata <- read.csv(file = '../data/healthcare-dataset-stroke-data-cleanbmi.csv')
head(strokedata)

age hypertension heart_disease ever_married work_type     avg_glucose_level
1 67  0            1             Yes          Private       228.69           
2 80  0            1             Yes          Private       105.92           
3 49  0            0             Yes          Private       171.23           
4 79  1            0             Yes          Self-employed 174.12           
5 81  0            0             Yes          Private       186.21           
6 74  1            1             Yes          Private        70.09           
  bmi  smoking_status  stroke
1 36.6 formerly smoked 1     
2 32.5 never smoked    1     
3 34.4 smokes          1     
4 24.0 never smoked    1     
5 29.0 formerly smoked 1     
6 27.4 never smoked    1

- 被清理的資料是符合以下條件的:
  - Gender為Other
  - BMI為N/A
  - Smoking status為Unknown
- 清理完的資料含有3425資料點

In [2]:
nrow(strokedata)

[1] 3425

<p style="page-break-before: always">

## Simple Models: 只含有一個解釋變數 

### 高血壓: 
$$
\log{\left(\frac{P[\rm{stroke}=1]}{P[\rm{stroke}=0]}\right)} = \beta_0 + \beta_{\rm{hypert}}x_{\rm{hypert}}
$$

In [3]:
model_hypert = glm(stroke ~ hypertension, data=strokedata, family=binomial(link="logit"))
summary(model_hypert)


Call:
glm(formula = stroke ~ hypertension, family = binomial(link = "logit"), 
    data = strokedata)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-0.5486  -0.2885  -0.2885  -0.2885   2.5298  

Coefficients:
             Estimate Std. Error z value Pr(>|z|)    
(Intercept)  -3.15821    0.09206 -34.305  < 2e-16 ***
hypertension  1.34048    0.16991   7.889 3.04e-15 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1410.9  on 3424  degrees of freedom
Residual deviance: 1358.1  on 3423  degrees of freedom
AIC: 1362.1

Number of Fisher Scoring iterations: 6


- <span style="color:blue"> $\beta$的z-test的p-value很小，高血壓與中風的預測是顯著的 </span>
- <span style="color:blue"> $\beta_{\rm{hypert}}=1.34 > 0$ : 表示有高血壓會增加中風的機率 </span>

### 心臟病
$$
\log{\left(\frac{P[\rm{stroke}=1]}{P[\rm{stroke}=0]}\right)} = \beta_0 + \beta_{\rm{heartd}}x_{\rm{heartd}}
$$

In [4]:
model_heartd = glm(stroke ~ heart_disease, data=strokedata, family=binomial(link="logit"))
summary(model_heartd)


Call:
glm(formula = stroke ~ heart_disease, family = binomial(link = "logit"), 
    data = strokedata)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-0.6198  -0.3025  -0.3025  -0.3025   2.4928  

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept)   -3.06125    0.08525 -35.908  < 2e-16 ***
heart_disease  1.50897    0.20231   7.459 8.73e-14 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1410.9  on 3424  degrees of freedom
Residual deviance: 1367.2  on 3423  degrees of freedom
AIC: 1371.2

Number of Fisher Scoring iterations: 5


- <span style="color:blue"> $\beta$的z-test的p-value很小，心臟病與中風的預測是顯著的 </span>
- <span style="color:blue"> $\beta_{\rm{heartd}}=1.509 > 0$ : 表示有心臟病會增加中風的機率 </span>

### Ever-Married(結婚有無)
$$
\log{\left(\frac{P[\rm{stroke}=1]}{P[\rm{stroke}=0]}\right)} = \beta_0 + \beta_{\rm{marry}}x_{\rm{marry}}
$$

In [5]:
model_marry = glm(stroke ~ ever_married, data=strokedata, family=binomial(link="logit"))
summary(model_marry)


Call:
glm(formula = stroke ~ ever_married, family = binomial(link = "logit"), 
    data = strokedata)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-0.3565  -0.3565  -0.3565  -0.2214   2.7279  

Coefficients:
                Estimate Std. Error z value Pr(>|z|)    
(Intercept)      -3.6964     0.2264  -16.33  < 2e-16 ***
ever_marriedYes   0.9722     0.2406    4.04 5.34e-05 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1410.9  on 3424  degrees of freedom
Residual deviance: 1390.4  on 3423  degrees of freedom
AIC: 1394.4

Number of Fisher Scoring iterations: 6


- <span style="color:blue"> $\beta_{\rm{marry}}=0.9722 > 0$ : 表示"有結過婚"會增加中風的機率 </span>
- <span style="color:blue"> 目前這裡的資料是含有"未達已婚年齡者的"，我們會在後面討論年齡與婚姻的對於中風預測的相互影響</span>

### Work-type(工作類型)
$$
\log{\left(\frac{P[\rm{stroke}=1]}{P[\rm{stroke}=0]}\right)} = \beta_0 + \beta_{\rm{GovJob}}x_{\rm{GovJob}} + \beta_{\rm{Neverworked}}x_{\rm{Neverworked}} + \beta_{\rm{Private}}x_{\rm{Private}} + \beta_{\rm{Selfemployed}}x_{\rm{Selfemployed}}
$$
- Baseline group: children

In [6]:
model_worktype = glm(stroke ~ work_type, data=strokedata, family=binomial(link="logit"))
summary(model_worktype)


Call:
glm(formula = stroke ~ work_type, family = binomial(link = "logit"), 
    data = strokedata)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-0.3985  -0.3188  -0.3188  -0.3188   2.4927  

Coefficients:
                         Estimate Std. Error z value Pr(>|z|)
(Intercept)            -1.757e+01  4.798e+02  -0.037    0.971
work_typeGovt_job       1.451e+01  4.798e+02   0.030    0.976
work_typeNever_worked  -1.793e-08  1.161e+03   0.000    1.000
work_typePrivate        1.461e+01  4.798e+02   0.030    0.976
work_typeSelf-employed  1.507e+01  4.798e+02   0.031    0.975

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1410.9  on 3424  degrees of freedom
Residual deviance: 1394.7  on 3420  degrees of freedom
AIC: 1404.7

Number of Fisher Scoring iterations: 16


- <span style="color:blue"> z-test的p-value都很大，表示我們沒辦法拒絕下面的$H_0$ </span>
\begin{align*}
&H_0: \beta_{\rm{GovJob}} = 0 \\
&H_0: \beta_{\rm{Neverworked}} = 0 \\ 
&H_0: \beta_{\rm{Private}} = 0 \\
&H_0: \beta_{\rm{Selfemployed}} = 0
\end{align*}

- <span style="color:blue"> 因此，工作型態對於預測中風是沒有幫助的，後面的模型不會考慮此解釋變數</span>

### Smoke Status(沒抽過煙、戒煙的人、正在抽煙的人)
$$
\log{\left(\frac{P[\rm{stroke}=1]}{P[\rm{stroke}=0]}\right)} = \beta_0 + \beta_{\rm{Former-smoke}}x_{\rm{Former-smoke}} + \beta_{\rm{smoke}}x_{\rm{smoke}}
$$
- Baseline group: 'never smoked'

In [7]:
strokedata$smoking_status <- factor(strokedata$smoking_status, levels= c('never smoked', 'formerly smoked', 'smokes'))
model_smoke = glm(stroke ~ smoking_status, data=strokedata, family=binomial(link="logit"))
summary(model_smoke)


Call:
glm(formula = stroke ~ smoking_status, family = binomial(link = "logit"), 
    data = strokedata)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-0.3758  -0.3297  -0.3047  -0.3047   2.4872  

Coefficients:
                              Estimate Std. Error z value Pr(>|z|)    
(Intercept)                    -3.0468     0.1117 -27.287   <2e-16 ***
smoking_statusformerly smoked   0.4318     0.1769   2.441   0.0146 *  
smoking_statussmokes            0.1621     0.1988   0.815   0.4149    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1410.9  on 3424  degrees of freedom
Residual deviance: 1405.1  on 3422  degrees of freedom
AIC: 1411.1

Number of Fisher Scoring iterations: 5


- <span style="color:blue">$\beta_{\rm{Former-smoke}}>0$, $\beta_{\rm{smoke}}>0$:  相比於不抽煙的人(Baseline)，有抽過煙會增加中風的機率</span>
- <span style="color:blue">$\beta_{\rm{Former-smoke}}>\beta_{\rm{smoke}}$: 這裡是比較違反直覺的地方，"戒煙的人"竟然比"目前還在抽煙的人"更容易中風</span>
- <span style="color:blue">$\beta_{\rm{smoke}}$ Z-test的p-value=0.4149: 表示抽煙狀態對於預測中風，沒有前面幾個變數來的重要</span>

## Model Selection: with/without smoke status

- <span style="color:blue"> 為了決定是否要將smoke status考慮進去最後的模型，我們用likelihood-ratio test做Model Selection</span>

- Full Model:
$$
\log{\left(\frac{P[\rm{stroke}=1]}{P[\rm{stroke}=0]}\right)} = \beta_0 + \beta_{\rm{hypert}}x_{\rm{hypert}}+ \beta_{\rm{heartd}}x_{\rm{heartd}}+ \beta_{\rm{marry}}x_{\rm{marry}} + \beta_{\rm{Former-smoke}}x_{\rm{Former-smoke}} + \beta_{\rm{smoke}}x_{\rm{smoke}}
$$

- Reduced Model: 
$$
\log{\left(\frac{P[\rm{stroke}=1]}{P[\rm{stroke}=0]}\right)} = \beta_0 + \beta_{\rm{hypert}}x_{\rm{hypert}}+ \beta_{\rm{heartd}}x_{\rm{heartd}}+ \beta_{\rm{marry}}x_{\rm{marry}}
$$

In [8]:
model_full = glm(stroke ~ hypertension + heart_disease + ever_married + smoking_status, data=strokedata, family=binomial(link="logit"))
summary(model_full)


Call:
glm(formula = stroke ~ hypertension + heart_disease + ever_married + 
    smoking_status, family = binomial(link = "logit"), data = strokedata)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-0.9457  -0.3195  -0.2799  -0.2228   2.8188  

Coefficients:
                              Estimate Std. Error z value Pr(>|z|)    
(Intercept)                    -3.9539     0.2399 -16.479  < 2e-16 ***
hypertension                    1.1381     0.1748   6.510 7.52e-11 ***
heart_disease                   1.2388     0.2099   5.902 3.60e-09 ***
ever_marriedYes                 0.7338     0.2451   2.994  0.00275 ** 
smoking_statusformerly smoked   0.2704     0.1824   1.482  0.13822    
smoking_statussmokes            0.0779     0.2036   0.383  0.70202    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1410.9  on 3424  degrees of freedom
Residual deviance: 1311.8  on 3419  degree

In [9]:
model_reduce = glm(stroke ~ hypertension + heart_disease + ever_married, data=strokedata, family=binomial(link="logit"))
summary(model_reduce)


Call:
glm(formula = stroke ~ hypertension + heart_disease + ever_married, 
    family = binomial(link = "logit"), data = strokedata)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-0.8899  -0.2930  -0.2930  -0.2013   2.7961  

Coefficients:
                Estimate Std. Error z value Pr(>|z|)    
(Intercept)      -3.8888     0.2308 -16.852  < 2e-16 ***
hypertension      1.1373     0.1746   6.513 7.37e-11 ***
heart_disease     1.2674     0.2085   6.080 1.20e-09 ***
ever_marriedYes   0.7621     0.2443   3.120  0.00181 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1410.9  on 3424  degrees of freedom
Residual deviance: 1313.9  on 3421  degrees of freedom
AIC: 1321.9

Number of Fisher Scoring iterations: 6


### Likelihood-Ratio Test
- Full Model與Reduced Model的差異可以轉換成  $H_0: \beta_{\rm{Former-smoke}} = \beta_{\rm{smoke}}= 0$

In [10]:
anova(model_reduce, model_full, test="LRT")

Resid. Df Resid. Dev Df Deviance Pr(>Chi) 
1 3421      1313.945   NA       NA        NA
2 3419      1311.770    2 2.175009 0.3370566

- 根據Deviance的p-value=0.3370
    - <span style="color:blue"> 不能拒絕 $H_0$，所以Final Model將不考慮"抽煙狀態"的變數</span>

## Final Model
$$
\log{\left(\frac{P[\rm{stroke}=1]}{P[\rm{stroke}=0]}\right)} = \beta_0 + \beta_{\rm{hypert}}x_{\rm{hypert}}+ \beta_{\rm{heartd}}x_{\rm{heartd}}+ \beta_{\rm{marry}}x_{\rm{marry}}
$$

In [11]:
summary(model_reduce)


Call:
glm(formula = stroke ~ hypertension + heart_disease + ever_married, 
    family = binomial(link = "logit"), data = strokedata)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-0.8899  -0.2930  -0.2930  -0.2013   2.7961  

Coefficients:
                Estimate Std. Error z value Pr(>|z|)    
(Intercept)      -3.8888     0.2308 -16.852  < 2e-16 ***
hypertension      1.1373     0.1746   6.513 7.37e-11 ***
heart_disease     1.2674     0.2085   6.080 1.20e-09 ***
ever_marriedYes   0.7621     0.2443   3.120  0.00181 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1410.9  on 3424  degrees of freedom
Residual deviance: 1313.9  on 3421  degrees of freedom
AIC: 1321.9

Number of Fisher Scoring iterations: 6


### 模型解釋:

由於這三個解釋變數都是類別型，我們可以用final model來探討不同的狀況
- 例如: 最左邊的藍色點，代表的是 "沒有結婚" "沒有高血壓" "沒有心臟病"的中風機率，可以看到非常的低。

```{image} ./images/cate_model_full_plot.png
:alt: glucose_stroke
:class: bg-primary mb-1
:width: 800px
:align: center
```

- <span style="color:blue"> $\beta_{\rm{heartd}} > \beta_{\rm{hypert}} > \beta_{\rm{marry}}>0$</span>
   - "有高血壓"、"有心臟疾病"與"有結過婚"都會增加中風機率。如上圖最右邊灰色的點，三種狀況都有，擁有最高的中風機率
   - 有心臟病的影響是最大的，其次是高血壓，最後是結婚狀態

<p style="page-break-before: always">